### PR TITLE
Analytics cookies are deleted and disabled when consent removed

### PR DESCRIPTION
**Pull request by vanitabarrett**
## What
There are a couple of related changes included in this PR:

- Make sure all code that deletes cookies sets the expiry to `Thu, 01 Jan 1970 00:00:00 GMT` rather than using `new Date()` which does not consistently work
- Move the above logic for cookie deletion to the `deleteCookie()` function, to be consistent
- Refactor `COOKIE_CATEGORIES` to lookup via category (e.g: 'analytics') rather than by cookie name (see individual commit for more details)
- Add `_gat_UA` cookies to the allow list
- Fixes to test setup to support the above changes: setting the testURL to a value other than localhost; making the logic which deletes cookies in the `afterEach` more generic so we don't need to manually update it each time a new cookie is added to the allow list.

## Why
Analytics cookies were being set to empty or blank values rather than being completely removed from the browser. This seemed to be related to us deleting the cookie using `new Date()` when we specify the domain, rather than manually setting it to a date in 1970.

Analytics cookies were also re-appearing after being deleted in some scenarios (see steps below). This was happening because we weren't disabling Google Analytics when we deleted the cookies, so it was re-creating them. We now set an object on the window to `true` or `false` depending on if the user has consented to analytics, following [Google's documentation](https://developers.google.com/analytics/devguides/collection/analyticsjs/user-opt-out)

Scenario where analytics cookies were reappearing (this should now be fixed):

1. Accept analytics cookies
2. Navigate to the cookies page and reject analytics cookies
3. Click a link to another page in the design system - as soon as the link is clicked, it triggers an analytics event (because analytics haven't been disabled) and re-creates the cookies

### CODE REVIEW

File: src/javascripts/components/cookie-functions.js [hunk 1]
```diff
@@ -246,5 +243,10 @@ function setCookie (name, value, options) {
 
 function deleteCookie (name) {
   document.cookie = name + '=; expires=Thu, 01 Jan 1970 00:00:00 GMT'
-  return null
+
+  // If the cookie still exists, let's try a more thorough way of deleting it,
+  // by specifyinh the domain and path
```
Comment (lfdebrux) [inline src/javascripts/components/cookie-functions.js:248]:
```suggestion
  // by specifying the domain and path
```

File: src/javascripts/components/cookie-functions.js [hunk 2]
```diff
@@ -161,11 +161,8 @@ export function resetCookies () {
     if (!options[cookieType]) {
       for (var cookie in COOKIE_CATEGORIES) {
         if (COOKIE_CATEGORIES[cookie] === cookieType) {
+          // Delete cookie
           Cookie(cookie, null)
```
Comment (lfdebrux) [inline src/javascripts/components/cookie-functions.js:165]: Any reason we can't just use `deleteCookie` here?

```suggestion
          deleteCookie(cookie)
```

**Review by lfdebrux — COMMENTED** [review]
Thanks for fixing this! Any idea why it wasn't picked up by the unit tests?

File: src/javascripts/components/cookie-functions.js (hunk 2 above)
Comment (vanitabarrett) [reply src/javascripts/components/cookie-functions.js:165]: @lfdebrux Technically, no, but we do (near the top of this file) suggest using the `Cookie()` helper function for getting, setting and deleting. I guess you could say that's more for files where you're importing cookie functions though...

Comment (lfdebrux) [reply src/javascripts/components/cookie-functions.js:165]: Yeah, I guess although we don't have to use the external API, it is best to be consistent. It does make me wonder whether those functions should then be encapsulated, to prevent usage, but that's for another day.

File: src/javascripts/components/cookie-functions.js [hunk 3]
```diff
@@ -246,5 +243,10 @@ function setCookie (name, value, options) {
 
 function deleteCookie (name) {
   document.cookie = name + '=; expires=Thu, 01 Jan 1970 00:00:00 GMT'
-  return null
+
+  // If the cookie still exists, let's try a more thorough way of deleting it,
+  // by specifying the domain and path
+  if (Cookie(name)) {
+    document.cookie = name + '=;expires=Thu, 01 Jan 1970 00:00:00 GMT;domain=' + window.location.hostname.replace(/^www\./, '.') + ';path=/'
+  }
```
Comment (lfdebrux) [inline src/javascripts/components/cookie-functions.js:251]: I'm not sure I understand why this is necessary. Can we add a unit test that covers this scenario?

File: src/javascripts/components/cookie-functions.js [hunk 4]
```diff
@@ -157,14 +159,19 @@ export function resetCookies () {
...
-          Cookie(cookie, null)
+      // Fetch that category lookup function
+      var cookiesInCategory = COOKIE_CATEGORIES[cookieType]
+      var cookies = document.cookie.split(';')
+
+      cookies.forEach(function (cookie) {
+        var cookieName = cookie.split('=')[0].replace(' ', '')
+
+        // Delete all cookies in that category
+        if (cookiesInCategory.indexOf(cookieName) >= 0) {
```
Comment (lfdebrux) [inline src/javascripts/components/cookie-functions.js]: I feel like the code getting the existing cookies is redundant; surely deleting a cookie that isn't already set should be fine, right? Or did you find problems with that in testing?

File: src/javascripts/components/cookie-functions.js [hunk 5]
```diff
@@ -237,7 +244,7 @@ function setCookie (name, value, options) {
     if (typeof options === 'undefined') {
       options = {}
     }
-    var cookieString = name + '=' + value + '; path=/'
+    var cookieString = name + '=' + value + ';path=/'
```
Comment (lfdebrux) [inline src/javascripts/components/cookie-functions.js:247]: Why is this change needed?

File: __tests__/cookie-functions.test.js [hunk 6]
```diff
@@ -15,8 +15,11 @@ describe('Cookie settings', () => {
 
   afterEach(() => {
     // Delete test cookies
-    document.cookie = '_ga=;expires=Thu, 01 Jan 1970 00:00:00 UTC'
-    document.cookie = 'design_system_cookies_policy=;expires=Thu, 01 Jan 1970 00:00:00 UTC'
+    var cookies = document.cookie.split(';')
+    cookies.forEach(function (cookie) {
+      var name = cookie.split('=')[0]
+      document.cookie = name + '=;expires=Thu, 01 Jan 1970 00:00:00 GMT;domain=' + window.location.hostname + ';path=/'
+    })
```
Comment (lfdebrux) [inline __tests__/cookie-functions.test.js:22]: 👍

File: src/javascripts/components/cookie-functions.js (hunk 5 above)
Comment (vanitabarrett) [reply src/javascripts/components/cookie-functions.js:247]: Oops, this wasn't intentional - will remove.

File: src/javascripts/components/cookie-functions.js (hunk 4 above)
Comment (vanitabarrett) [reply src/javascripts/components/cookie-functions.js]: That's a really good shout, for some reason it didn't occur to me 👍 I don't think that would cause any problems... I'll do some testing to double check

Comment (vanitabarrett) [reply src/javascripts/components/cookie-functions.js]: @lfdebrux I've tweaked this now so we no longer fetch the existing cookies ([commit](https://github.com/alphagov/govuk-design-system/pull/1791/commits/7318988fe0ca0045746ee2528de5713d76b8d1d2#diff-54c970228ba5fe3657fcdee2af4308d31ce1c41eff42a21b9405e919664cd25fR164))

**Review by lfdebrux — APPROVED** [review]
Amazing work getting to the bottom of all this! 🙌 

Works for me, feels like we're almost there 😅
